rfcs/2022--01-06-8596-tcp-source-backpressure.md

```diff
@@ -0,0 +1,160 @@
+`x events + y requests`. This would attempt to limit the in-flight events to `x`, plus `y` full requests of events.
+
+The main drawback to this approach is that you have to be able to accept a request before you know how many events
+are in that request, and the total in-flight count can't be updated until after the decoding has finished. A concurrency
+limit must be placed on request decoding (equal to `y` above) in order to limit the overall requests in-flight.
+
+This leads us to the question of what should the value of `x` and `y` be? `x` (the number of events in-flight) can probably
+be set to 0 and ignored, since we already have event buffering in other parts of the system (the pipeline and sink buffers).
+Choosing `y` is balancing performance with memory usage.
+
+I propose that `y` is a value dynamically determined by the request size of previous requests.
+The "request size" will simply be defined as the number of events contained in a request. The running-average of
+these will be used to estimate the size of future requests. Now a target number of in-flights requests can be chosen.
+`y` will be set to a value such that actual in-flight messages + estimated in-flight messages is less than the target.
+Once a request is decoded the actual in-flight messages can be increased, and as they are flushed into the pipeline
+the count will be decreased. The target value will be a constant value somewhat arbitrarily chosen based on perf tests.
+The exact value will be proposed in the final implementation.
+
+A dynamic value for `y` is useful because the size of the requests can vary drastically depending on the source / protocol used.
+Each request may only contain a single event, or as noted above can contain tens of thousands. Using a constant
+value would likely penalize sources using small requests by artificially limiting concurrency, or allowing
+sources with large requests to have too many events in memory.
+
+On top of the dynamic value for `y`, some hard limits will be added. Regardless of the current request size estimate,
+1 request must always be allowed to be processed when the actual in-flight message count is zero to ensure forward
+progress can be made. In addition, an upper-bound will be set equal to the number of CPU's on the system. A
+value higher than this has no benefit, but would use more memory.
+
+### Rationale
+
+- No user facing config needed, this will work by default.
+
+### Drawbacks
+
+- More complicated to implement than option 1.
+- Requires choosing an appropriate "target" value. If too low, it could limit overall throughput. If too high, it could use too much memory.
+
+## Proposal
+
+I am proposing implementing option 3 (dynamic request limit).
+This will limit the number of requests being processed concurrently to balance memory usage with performance.
+I don't think option 2 (dynamic connection limit) is feasible to implement, and option 1 (static connection limit) is
```

Comment:
Would we also back out the existing (optional) config for Option 1, or will that remain available for users to configure?

Reply:
I was planning on leaving that as an option.

Reply:
Sorta mirrors the hard coding instead of ARC, though it's connections vs requests here 👍